### ISSUE TITLE
[enhancement] Using PyCOMPSs as the Wrapper Engine

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1176]
In GitLab by @manuel-g-castro on Dec 4, 2023, 16:49

Hello, @dbeltrankyl, @kinow, @mcastril, and @ macosta:

I am looking into ways of optimizing the wrapper engine, that is the execution of the tasks within the same package, and one of the options is to use COMPSs runtime, throught the library PyCOMPSs. There are plenty of features where we can see benefits of using: 

- Abstract away from the memory. Tasks within COMPSs see the whole allocated memory in across all nodes as a single pool. The Runtime takes care of the data locality and minimizing the data movement
- We could set a fixed amount of resources, and the Runtime will take care of executing tasks whenever it is possible. (e.g. two tasks need a full node. Instead of asking for 2 nodes to run them in parallel, we could ask for a single one and the Runtime takes care of executing the tasks whenever the node is free)
- We can define tasks as being binary, hence we don't need to change the underlying code of the scripts/models. Daniele has confirmed us that COMPSs doesn't really care what is being run.

**Applicability**

We had a meeting with Daniele Lezzi, from Computer Science's Workflows group, where we layed down the example of Auto-ECEarth3. In this workflow, we have one post-process job per submodel (ocean, ifs, lpjm, etc) and each of them have different computing needs (check the runtime in the tree view). A first proof of concept would be to run all of these tasks on a single wrapper, like in the graph view below.

**The challenges**

_I am concerned about the portability of the tool_. Daniele told us that is possible to run COMPSs runtime in a containerized way, and even (as far as I understand) from outside of remote, if the platform does allow for node connectivity with the outside world. For MareNostrum, they have everything setup to be easy executing using their launcher from Remote. Although, in our case, that would mean to abdicate Autosubmit's allocation of SLURM via Sbatch. That is something that, as of now, is not beneficial for Autosubmit. 

Also, we want to make it generic, in the sense that Autosubmit creates the PyCOMPSs workflow at the request of the user, and that seems a pretty substantial problem.

Miguel said that Javier Conejero tried to do a copy of one of our Workflows (was it auto-ecearth3?) in PyCOMPSs. But that got eventually stopped.

**Next Steps**

- I plan to test PyCOMPSs with a couple of simple tests to understand it. 

- Then, if it is proven beneficial to follow with tool, I will try to abstract away and make it generic.

![postprocess-tree](https://earth.bsc.es/gitlab/es/autosubmit/uploads/c6737437d34c30f11bccf560c90abc4c/postprocess-tree.png)

![postprocess-ecearth-wrapper](https://earth.bsc.es/gitlab/es/autosubmit/uploads/640b232462a9ee5e473fc166b261b235/postprocess-ecearth-wrapper.png)

**Comment by kinow** [comment]
In GitLab by @ macosta on Dec 4, 2023, 18:44

@manuel-g-castro one point that it could be useful.

PyCOMPSs has portability problems as we know and some specific limitations, though I know that from the research and possibilities point of view is great.

I think we should work on a "homemade" solution to optimize the wrapper engine and additionally create a test case where our solution and PyCOMSs can be compared.

Why this proposal?

We should not evaluate if PyCOMSs could be used for our projects and operational stuff for now. We should evaluate if PyCOMSs can do it better than our solutions in a controlled environment. 

I think you can start with the simple tests as you mention and discuss more in detail in our meeting next Friday.

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Jan 30, 2024, 20:42

changed due date to February 29, 2024

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Feb 7, 2024, 11:29

In a private conversation with a colleague, he referenced a couple of issues [1](https://github.com/bsc-wdc/compss/issues/9) and [2](https://github.com/bsc-wdc/tutorial_apps/issues/2) to help install locally COMPSs.

And also mentioned that COMPSs is not easily installed in remote platforms. He had encountered some hard coded configurations of the cluster network. Apparently COMPSs always assumed that the host from infinity band is lb0, but in this particular machine it is a different one.
  
He referenced that Suvi, from Alfred Wegener Institute (AWI), has a newer example, if required.

Moreover, Deniele Lezzi from here at the BSC contacted back making himself available.

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Feb 7, 2024, 11:30

I will now try to run some tutorials locally, to then move to MareNostrum 4. If everything goes smoothly I will try to make a first group of tasks from auto-ecEarth3 run using it.

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Feb 7, 2024, 17:59

First day working with this and it was a major failure. In [the tutorial](https://github.com/bsc-wdc/tutorial_apps) that I am following they ask for a system-wide installation of pycompss by doing a `sudo pip3 install pycompss-cli`. I don't want to do that to avoid potentially breaking something in my system, hence I am doing a `venv`. 

Well. I failed. I created a venv by issuing a `python3 -m venv pycompss` in my virtual environments folder. Installed pycompss-cli via `pip3` once activated the environment, plus installed `wheels`, because it was complaining about the lack of it (error `error: invalid command 'bdist_wheel'`), and `docker`, as stated by [this issue](https://github.com/bsc-wdc/tutorial_apps/issues/2) opened by Bruno. With that done, I had `pycompss-cli` working. But, then it was time of downloading the docker image and actually start running it. Well, after grabbing a coffee while waiting for the download (hogging most of the wifi, peaks of 1200 pckgs/s), I had the container working. At least that is what showed in the `docker container ls`.

Now, I should be able to run `pycompss monitor start` and `pycompss jupyter`. The first starts, and hangs silently, while simultanously opening a firefox (it's hardcoded for it...) tab with `localhost:8080/compss-monitor`. Nothing loads. Issuing a `curl localhost:8080` I get `curl: (56) Recv failure: Connection reset by peer`. The latter is even worst, where I get the following:

```
Traceback (most recent call last):
  File "/home/mgimenez/Documents/local/dev/venvs/pycompss/bin/pycompss", line 11, in <module>
    load_entry_point('pycompss-cli==3.3.2', 'console_scripts', 'pycompss')()
  File "/home/mgimenez/Documents/local/dev/venvs/pycompss/lib/python3.8/site-packages/pycompss_cli/cli/pycompss.py", line 45, in main
    ActionsDispatcher().run_action(arguments)
  File "/home/mgimenez/Documents/local/dev/venvs/pycompss/lib/python3.8/site-packages/pycompss_cli/core/actions_dispatcher.py", line 53, in run_action
    action_func()
  File "/home/mgimenez/Documents/local/dev/venvs/pycompss/lib/python3.8/site-packages/pycompss_cli/core/local/actions.py", line 111, in jupyter
    if jupyter_args[0] == 'lab':
IndexError: list index out of range
```

I tried utilizing the `latest` image instead of the `2.1` used in the tutorial. Didn't make a difference, and it had me take a second coffee waiting for it. Makes sense, since this seems like an issue related to launch script or the jupyter version or something.

I tried to install `jupyter-notebook` in the virtual environment by doing a `pip3 install jupyterlab`, even though I have it system-wide. It didn't work.

Now I am wondering if I should do a system-wide installation (yucky) or see if there is a newer version of `pycompss-cli` which is working. Good for me is that Rohan did this training, then he should be able to help me out (aslong as we keep my laptop sanitized of pip installations, really trying to keep everything tidy here).

**Comment by kinow** [comment]
FWIW, I also had tried with a venv, and wasn't really keen on using `sudo pip` as I have no idea what it would install/configure. But I think I didn't get the IndexError (can't really recall what was my error).

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Feb 8, 2024, 17:25

Okay, second day: managed to run one of the examples.

The problem that I was facing is that: after running the `pycompss init (...)` you have to run a `pycompss env change docker-tutorial`. That is because, otherwise, the cli will try to run on the default environment, which will have no access to the just initiated docker container.

Another issue that I faced is that the init automatically binds folder from where the command issued. Since I was not running the command from the same directory as where the notebooks where, I wasn't able to run them. Obviously this applied only if you run from the docker image :smile: .

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Feb 9, 2024, 11:01

mentioned in issue mgimenez/phd-project-management#4

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Feb 19, 2024, 11:17

mentioned in issue mgimenez/phd-project-management#11

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Feb 22, 2024, 14:57

So from today's meeting with Deniele, Javier, Raül, Mario, and Gladys we have:

- RO-Crates: on top of his mind, Raül proposes to just parse the important bits of COMPSs and then afterwards include that on the RO-crate object from Autosubmit. All of that depends on the granularity of the information we aim for. Correct me if I am wrong, @kinow, but there are plenty of levels of verbosity we can create valid RO-crates from. Maybe just match whatever COMPSs offers on Autosubmit? Hopefully there is no mismatch.
- COMPSs: Jorge showed me the launch script he has on MareNostrum. Basically, there is a command `enqueue_compss`, at mn4, which is reponsible of finding out who are the nodes which the runtime will manage and who is the master. Then he lifts the runtime. This is done prior to launching pycompss with the script which has the workflow. Now the idea is to create an autosubmit slurm header for compss with these initial steps, and then launch compss with the workflow definition. 

I will try to do by hand Jorge's proposal. Although I am missing the scripts...

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Feb 26, 2024, 17:17

Okay. Since Thursday I have been trying to figure out how I will develop something with COMPSs. The current way that they provide is by giving: a pip installable script and a docker image. The docker image contains COMPSs per-se, the other script is just for managing it (starting the container with the correct bindings etc). Problem is: I want to develop an ece2cmor wrapper, which I need to include whithin this docker-image (compss/compss-tutorial). ece2cmor is included with a micromamba environment, so I need to configure a micromamba environment in this docker image. 

This last part has taken me from Friday up until now. I am working on a new docker image pulling from a base COMPSs and a base micromamba, where I install ece2cmor. 

As frustrating as it has been this past couple of days, I think (and talking with Bruno) this is the best away to avoid having random issues as time passes by. Using a single docker image we make sure that both dependencies are met: COMPSs and ece2cmor.

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Feb 28, 2024, 10:38

Well. I have decided to merge both Dockerimages onto a single one: compss/compss-tutorial and micromamba/micromamba:1.5.6. In this single image, I install ece2cmor's environment and activate it. 

I have found an annoying issue: that we have to be extra careful to when and how the environment is loaded within the container. In my case, I have [found](https://micromamba-docker.readthedocs.io/en/latest/advanced_usage.html#on-docker-exec) that during `docker exec ...` the environment is not loaded. Hence, it is necessary to change the path environment variable in the Dockerfile `ENV PATH "$MAMBA_ROOT_PREFIX/bin:$PATH"  # Not a preferred method!`. 

Now I have a working image where I can execute ece2cmor through COMPSs. Finally, after 2 weeks or so I can start developing something.

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Mar 5, 2024, 19:30

changed due date to March 31, 2024

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Mar 5, 2024, 19:42

Update: my solution is a mess. Luckily for me, it is just a development environment. 

After successfully creating an image with both compss and ece2cmor, so I can run the later in the first, I had an annoying error saying that it could not import `compss` when running `runcompss` within the container. 

The issue was a juxtaposition of two issues (kinda of like a very nasty hydra):

- micromamba overwrites `PYTHONPATH` variable. The reason is obvious: it is a virtual environment. Hence it should not fetch software outside of it (which is exactly the type venv sacrilegy that I want to do). The workaround (again, despicable and disgusting) is to run `export PYTHONPATH="/opt/COMPSs/Bindings/python/3"`
- with the proper (horrible) path set, we have to match the compiled version of compss with the one executing python. Because within the environment the default version of `python3` was `3.11.10`, it was looking for libraries not available. COMPSs is only available with 3.8 as explictly shown in the file name `compss.cpython-38-x86_64-linux-gnu.so` (check the cpython-38). So I changed the `environment.yml` of ece2cmor to ask for `python==3.8.10`.

Now I am able to execute `runcompss` with the COMPSs workflow. Still failing though, but that is probably because compss wants to do ssh connections to localhost.

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Mar 6, 2024, 13:03

Good news! I got the COMPSs workflow to work correctly! 

I made new changes to the Frankenstein-ish image container to set `PYTHONPATH` environment variable with COMPSs bindings and it is working. Now, out of the box, we have both COMPSs and ece2cmor.

Another issue I had to solve was that the docker image needs to be properly setup for the runtime to work. This is handled by `pycompss-cli`, which I was bypassing because I assumed it was unnecessary. Well, it is not. The runtime needs something from there. So, what I do, is a let pycompss-cli to run the container. Then I attach a session on this container, properly set up by `pycompss-cli`.

All of that is for the development environment, which is quite frankly a mess. 

As for actually running the ece2cmor I found a couple of issues:

1) the decorator with the arguments for the binary was wrongly implemented. Now I changed it for, where the double brackets will be subtituted by the functions parameters. This looks interesting since I think it can be easily created on the fly for any binary. 

```
@binary(binary="ece2cmor", args= "--exp {{expid}} {{model}} --meta {{meta}} --varlist {{varlist}} --odir {{odir}} {{input_dir}}")
@ task()
def ece2cmor(input_dir, model, expid, meta, varlist, odir):
    pass

```

2) Within COMPSs runtime, the micromamba environment was not being loaded. COMPSs developers thought about that and included a flag `--env_script=/home/user/activate.sh` where you specify the file to be sourced by the workers. So I had to include micromamba's base environment path so that ece2cmor is available for the workers

```
export PATH=/opt/conda/bin:/opt/conda/condabin:/opt/conda/bin:/opt/conda/bin:/opt/conda/condabin:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/COMPSs/Runtime/scripts/user:/opt/COMPSs/Bindings/c/bin:/opt/COMPSs/Runtime/scripts/utils
```

**Comment by kinow** [comment]
Did you get the right binary command line in the end? I couldn't find docs explaining where I could use `{{expid}}` or not (`mpi` has some arguments that do not support it, while others do :confused: )

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Mar 6, 2024, 14:20

Yess! I used [this page's](https://compss-doc.readthedocs.io/en/stable/Sections/02_App_Development/02_Python/01_1_Task_definition/Sections/06_Other_task_types/01_Binary_decorator.html?highlight=binary#binary-decorator) code 67 as a reference. You have to match the bracket's variable name with the argument in the function and it worked for me. As for MPI, I do not know...

```
from pycompss.api.task import task
from pycompss.api.binary import binary
from pycompss.api.parameter import *


@binary(binary="date", args= "-d {{param_1}}")
@ task()
def print_date(param_1):
     pass

if __name__=='__main__':
    print_date("next Monday")
```

**Comment by kinow** [comment]
>As for MPI, I do not know...

We spent quite a while trying to understand why some values were not being replaced in our workflow, although in our case we were already launching things with Slurm, without tests at that time... so every time we tried it, we spent ~10 minutes waiting for the job to be queued, executed, and searching the logs (definitely not fun, not recommended): https://github.com/bsc-wdc/compss/issues/10#issuecomment-1959214618

So in case you use `@mpi`, watch out for where you try to use `{{var}}`. (I don't know if it works everywhere, for every variable…)

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Apr 16, 2024, 12:26

mentioned in issue mgimenez/phd-project-management#10

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Apr 24, 2024, 10:37

mentioned in issue mgimenez/phd-project-management#18

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Apr 24, 2024, 16:19

changed due date to May 31, 2024

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on May 21, 2024, 11:38

Last week Miguel suggested for me to test the sample script of COMPSs on MN5, just to make sure that everything is working. Well, I have ran into issues and [I have already contacted the developers](https://github.com/bsc-wdc/compss/issues/14). 

Basically, rerunning the script within an interactive session fails to some module errors.

**Comment by kinow** [comment]
In GitLab by @manuel-g-castro on Aug 2, 2024, 15:56

moved to mgimenez/phd-project-management#33
